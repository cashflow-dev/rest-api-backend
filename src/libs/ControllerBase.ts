/* eslint-disable no-param-reassign */
import { ParameterizedContext } from 'koa';
import { serviceLaunch } from '../services';
import { SuccessBody } from '../interfaces/SuccessBody';
import { handleHttpErrors } from './utils';
import HTTP_STATUS_CODE from '../enums/HTTP_STATUS_CODE';

const ControllerBase = (service: any) => {
  const controller: any = {};

  controller.create = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'create', context);
      context.status = HTTP_STATUS_CODE.CREATED;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.find = async (context: ParameterizedContext): Promise<void> => {
    try {
      const { users, next } = await serviceLaunch(service, 'find', context);
      context.staus = HTTP_STATUS_CODE.OK;
      const body: SuccessBody = {
        message: 'ok',
        data: users,
        next,
        statusCode: HTTP_STATUS_CODE.OK,
      };
      context.body = body;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.findById = async (context: ParameterizedContext): Promise<void> => {
    try {
      const user = await serviceLaunch(service, 'findById', context);
      context.staus = HTTP_STATUS_CODE.OK;
      const body: SuccessBody = {
        message: 'ok',
        data: user,
        statusCode: HTTP_STATUS_CODE.OK,
      };
      context.body = body;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.update = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'update', context);
      context.status = HTTP_STATUS_CODE.NO_CONTENT;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.deleteById = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'delete', context);
      context.status = HTTP_STATUS_CODE.NO_CONTENT;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  return controller;
};

export default ControllerBase;
