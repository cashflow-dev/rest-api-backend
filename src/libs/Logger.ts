import moment from 'moment';
import bunyan from 'bunyan';
import chalk from 'chalk';
import 'dotenv/config';

const log = bunyan.createLogger({
  name: `${process.env.APP_NAME}`,
  src: true,
  streams: [
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'info',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'warn',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'error',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'fatal',
    },
  ],
});

class Logger {
  public static info(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.inverse.bold(logMessage));
    log.info(logMessage);
  }

  public static warn(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgYellow.black.bold(logMessage));
    log.warn(logMessage);
  }

  public static error(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgRed.bold.white(logMessage));
    log.error(logMessage);
  }

  public static vebose(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    log.info(logMessage);
  }

  public static debug(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    log.debug(logMessage);
  }

  public static fatal(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgRed.bold.white(logMessage));
    log.fatal(logMessage);
  }
}

export default Logger;
