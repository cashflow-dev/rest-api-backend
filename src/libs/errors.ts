export class NotFoundError extends Error {
  public constructor() {
    super('NotFoundError');
    Error.captureStackTrace(this, NotFoundError);
  }
}

export class UnauthorizedError extends Error {
  public constructor() {
    super('Unauthorized');
    Error.captureStackTrace(this, UnauthorizedError);
  }
}

export class ValidationError extends Error {
  public validationMessages: string[];

  public constructor(validationMessages: string[]) {
    super('ValidationError');
    this.validationMessages = validationMessages;
    Error.captureStackTrace(this, ValidationError);
  }
}
