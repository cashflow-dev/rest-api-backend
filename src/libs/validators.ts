import Joi from '@hapi/joi';
import joiObjectId from 'joi-objectid';
import _ from 'lodash';
import { InputData } from '../interfaces/InputData';
import ValidationError from '../helpers/ValidationError';

const ObjectId = joiObjectId(Joi);

export const composeValidators = (...functions: ((inputData: InputData) => void)[]) => {
  return function(inputData: InputData) {
    functions.forEach(func => {
      func(inputData);
    });
  };
};

export const validateId = (inputData: InputData) => {
  const schema = {
    params: {
      id: ObjectId()
        .required()
        .error(() => 'Not a valid id'),
    },
  };

  const paramsSchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, paramsSchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};

export const validateSearch = (inputData: InputData) => {
  const schema = {
    query: {
      search: Joi.string(),
    },
  };

  const querySchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, querySchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};
export const validateQuery = (inputData: InputData) => {
  const schema = {
    query: {
      next: ObjectId().error(() => 'Not a valid cursor'),
      fields: Joi.string(),
      limit: Joi.number().max(50),
    },
  };

  const querySchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, querySchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};
