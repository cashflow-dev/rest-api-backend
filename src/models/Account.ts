import { Schema, model, Document, Types } from 'mongoose';
import { timestamps } from 'mongoose-timestamp';

const { ObjectId } = Schema.Types;

interface Account extends Document {
  ownerId: Types.ObjectId;
  name: string;
  balance: number;
}

const accountSchema = new Schema({
  ownerId: ObjectId,
  name: String,
  balance: Number,
});

accountSchema.plugin(timestamps);

export default model<Account>('Account', accountSchema);
