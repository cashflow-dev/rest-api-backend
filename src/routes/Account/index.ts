import Router from 'koa-router';
import AccountController from '../../controllers/account';

const router = new Router({ prefix: '/account' });
const controller = AccountController();

/**
 *
 * @api {get} /api/v1/account GET MANY
 * @apiVersion 1.0.0
 * @apiName getAccounts
 * @apiGroup Account
 *
 * @apiDescription Get Accounts
 * Default limit: 50,
 * Max limit: 50
 * Supports pagination with the use of a cursor (next) field
 *
 * @apiExample {curl} Example usage:
 *     curl -v -X GET http://localhost:3003/api/v1/account
 *
 * @apiExample {curl} Example usage - Cusom limit:
 *     curl -v -X GET http://localhost:3003/api/v1/account?limit=2
 *
 * @apiExample {curl} Example usage - Partial:
 *     curl -v -X GET http://localhost:3003/api/v1/account?fields=email,_id
 *
 * @apiExample {curl} Example usage - Filter:
 *     curl -v -X GET http://localhost:3030/api/v1/account?email=demo@demo.com
 *
 * @apiSuccess (200) {Array} data Array of Account Objects
 * @apiSuccess (200) {String} next Cursor for next document in pagination
 * @apiError (Error responses) {String} message Internal Server error
 * @apiError (Error responses) {String} errorCode Code to find more information in the docs.
 * @apiError (Error responses) {String} errors Array of error messages.
 *
 * @apiSuccessExample {JSON} Success-Response:
 * {
 *    "message": "ok",
 *    "data": [
 *        {
 *            "_id": "5d40ad6f53e62706059c64ad",
 *            "balance": 100
 *        },
 *        {
 *            "_id": "5d40ad37fc77de0422109374",
 *            "balance": 100
 *        },
 *        {
 *            "_id": "5d3e17bb056dc3672600971e",
 *            "balance": 100
 *        }
 *    ],
 *    "next": "5d3e17bb056dc3672600971e",
 *    "statusCode": 200
 * }
 *
 *
 * @apiErrorExample {JSON} Error-Response:
 *  {
 *    "statusCode": 503,
 *    "errorCode": 503,
 *    "errors": [
 *        "Could not connect to database"
 *    ],
 *    "message": "ServiceUnavailable"
 * }
 */
router.get('/', controller.find);
/**
 *
 * @api {get} /api/v1/account/:id GET BY ID
 * @apiVersion 1.0.0
 * @apiName getAccount
 * @apiGroup Account
 *
 * @apiExample {curl} Example usage:
 *     curl -X GET 'http://localhost:3003/account/5d3e14c3056dc3672600971d?fields=balance,name,_id' -H 'Content-Type: application/json'
 *
 * @apiSuccess (200) {data} User Object
 * @apiError (Error responses) {String} message ValidationError
 * @apiError (Error responses) {String} errorCode Code to find more information in the docs.
 * @apiError (Error responses) {String} errors Array of error messages.
 *
 * @apiSuccessExample {JSON} Success-Response:
 * {
 *    "message": "ok",
 *    "data": {
 *        "_id": "5d3e17bb056dc3672600971e",
 *        "name": "test",
 *        "balance": 100
 *    },
 *    "statusCode": 200
 * }
 *
 * @apiErrorExample {JSON} Error-Response:
 *  {
 *    "statusCode": 400,
 *    "errorCode": 400,
 *    "errors": [
 *        "Not a valid id"
 *    ],
 *    "message": "ValidationError"
 * }
 */
router.get('/:id', controller.findById);
/**
 *
 * @api {patch} /api/v1/account/:id UPDATE
 * @apiVersion 1.0.0
 * @apiName updateAccount
 * @apiGroup Account
 *
 * @apiExample {curl} Example usage:
 *     curl -X PATCH http://localhost:3003/account/5d3e14c3056dc3672600971d -H 'Content-Type: application/json'  -d '{"balance": 200}'
 *
 * @apiSuccess (204) {Empty} Empty No content
 * @apiError (Error responses) {String} message ValidationError
 * @apiError (Error responses) {String} errorCode Code to find more information in the docs.
 * @apiError (Error responses) {String} errors Array of error messages.
 *
 * @apiSuccessExample {JSON} Success-Response:
 * status: 204
 *
 * @apiErrorExample {JSON} Error-Response:
 *  {
 *    "statusCode": 400,
 *    "errorCode": 400,
 *    "errors": [
 *        "Not a valid id"
 *    ],
 *    "message": "ValidationError"
 * }
 */
router.patch('/:id', controller.update);
/**
 *
 * @api {delete} /api/v1/account/:id DELETE
 * @apiVersion 1.0.0
 * @apiName deleteAccount
 * @apiGroup Account
 *
 * @apiExample {curl} Example usage:
 *     curl -X DELETE http://localhost:3003/account/5d3e17bb056dc3672600971e -H 'Content-Type: application/json' \
 *
 * @apiSuccess (204) {Empty} Empty No content
 * @apiError (Error responses) {String} message ValidationError
 * @apiError (Error responses) {String} errorCode Code to find more information in the docs.
 * @apiError (Error responses) {String} errors Array of error messages.
 *
 * @apiSuccessExample {JSON} Success-Response:
 * status: 204
 *
 * @apiErrorExample {JSON} Error-Response:
 *  {
 *    "statusCode": 400,
 *    "errorCode": 400,
 *    "errors": [
 *        "Not a valid id"
 *    ],
 *    "message": "ValidationError"
 * }
 */
router.delete('/:id', controller.deleteById);
/**
 *
 * @api {post} /api/v1/account CREATE
 * @apiVersion 1.0.0
 * @apiName createAccount
 * @apiGroup Account
 *
 * @apiExample {curl} Example usage:
 *     curl -X POST http://localhost:3003/account -H 'Content-Type: application/json' -d '{ "ownerId": "5d33c18dc1629243f43db274", "name": "test", "balance": 100 }'
 *
 * @apiSuccess (201) {Empty} Created
 * @apiError (Error responses) {String} message ValidationError
 * @apiError (Error responses) {String} errorCode Code to find more information in the docs.
 * @apiError (Error responses) {String} errors Array of error messages.
 *
 * @apiSuccessExample {JSON} Success-Response:
 * status: 201 Created
 *
 * @apiErrorExample {JSON} Error-Response:
 *  {
 *    "statusCode": 400,
 *    "errorCode": 400,
 *    "errors": [
 *        "ownerId is required"
 *    ],
 *    "message": "ValidationError"
 *}
 */
router.post('/', controller.create);
export default router;
