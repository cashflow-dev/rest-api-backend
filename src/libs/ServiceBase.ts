import _ from 'lodash';
import { InputData } from '../interfaces/InputData';
import { Service } from '../interfaces/Service';
import { composeValidators, validateId, validateQuery } from './validators';
import { NotFoundError } from './errors';
import { getRequestedFields, handleMongoErrors } from './utils';

export class ServiceBase implements Service {
  public input: InputData;

  public DEFAULT_LIMIT: number;

  public Model: any;

  public requestedFields: any;

  public queryFilter: any;

  public validators: any;

  public constructor(input: InputData) {
    this.DEFAULT_LIMIT = 50;
    this.input = input;
    this.validators = {
      find: composeValidators(validateQuery),
      findById: composeValidators(validateId),
      delete: composeValidators(validateId),
    };

    /* eslint-disable no-underscore-dangle */
    this.requestedFields = this.input.fields ? getRequestedFields(this.input.fields) : { _id: 1 };
    this.queryFilter = this.input.next
      ? {
          _id: { $lt: this.input.next },
        }
      : {};
  }

  public async findById(): Promise<any> {
    let result;
    try {
      result = await this.Model.findById(this.input.params.id).select(this.requestedFields);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }

  public async create(): Promise<any> {
    let result;
    try {
      const model = new this.Model(this.input.body);
      result = await model.save();
    } catch (e) {
      handleMongoErrors(e);
    }
    return result;
  }

  public async find(): Promise<any> {
    let results: any;
    try {
      results = await this.Model.find(this.queryFilter)
        .select(this.requestedFields)
        .sort({ _id: -1 })
        .limit(Number(this.input.limit) || this.DEFAULT_LIMIT);
    } catch (e) {
      handleMongoErrors(e);
    }
    return { users: results, next: _.get(results[results.length - 1], '_id', null) };
  }

  public async update(): Promise<any> {
    let result = null;
    try {
      result = await this.Model.findByIdAndUpdate(this.input.params.id, this.input.body);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }

  public async delete(): Promise<any> {
    let result = null;
    try {
      result = await this.Model.findByIdAndDelete(this.input.params.id);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }
}

export default ServiceBase;
