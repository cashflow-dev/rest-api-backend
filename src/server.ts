/* eslint-disable import/prefer-default-export */
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import logger from 'koa-logger';
// TODO: See issue #19
// import memwatch from 'node-memwatch';
import cors from '@koa/cors';
import fs from 'fs';
import router from './routes/index';
import { connect } from './models';
import Logger from './libs/Logger';
import 'dotenv/config';

const LOG_PATH = `${__dirname}/${process.env.LOG_PATH}`;

if (!fs.existsSync(LOG_PATH)) {
  fs.closeSync(fs.openSync(LOG_PATH, 'w'));
}

// TODO: See issue #19
// memwatch.on('leak', Logger.fatal);
// memwatch.on('stats', Logger.fatal);

const DATABASE = `${process.env.MONGODB_URI}`;

const PORT = 3003;

connect(
  DATABASE,
  Logger.fatal,
);
export const app = new Koa();
app.use(cors());
app.use(bodyParser());
app.use(logger((str, args) => Logger.debug(`${str}: ${args}`)));
app.use(router());

if (!module.parent) {
  app.listen(PORT, (): void => {
    Logger.debug(`API Server listening on port: ${PORT}`);
  });
}
