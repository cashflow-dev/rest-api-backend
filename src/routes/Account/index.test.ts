/* eslint-disable no-underscore-dangle */
import ObjectID from 'bson-objectid';
import request from 'supertest';
import { app } from '../../server';
import Account from '../../models/Account';
import { stripIdAndNext } from '../../libs/utils';

let server: any;
describe('/account', () => {
  let accountId: any;
  beforeAll(async () => {
    server = app.listen();

    const promises = [];
    for (let i = 0; i < 5; i += 1) {
      const account = new Account({ balance: i, name: `${i}`, ownerId: ObjectID.generate() });
      promises.push(account.save());
    }
    const response = await Promise.all(promises);
    accountId = response[0]._id;
  });
  afterAll(async () => {
    await server.close();
  });
  describe('GET /account', () => {
    it('should match snapshot', async () => {
      const response = await request(server).get('/account?fields=balance,name');
      expect(stripIdAndNext(response.text)).toMatchSnapshot();
      expect(response.status).toBe(200);
      expect(JSON.parse(response.text).data.length).toBeGreaterThan(0);
    });
  });
  describe('GET /account/:id', () => {
    it('should match snapshot', async () => {
      const response = await request(server).get(`/account/${accountId}`);
      expect(stripIdAndNext(response.text)).toMatchSnapshot();
      expect(response.status).toBe(200);
    });
  });
  describe('POST /account', () => {
    describe('with all fields', () => {
      const account = { balance: 200, name: 'test', ownerId: ObjectID.generate() };
      let response: any;
      let dbAccount: any;

      beforeAll(async () => {
        response = await request(server)
          .post('/account')
          .send(account);
        dbAccount = await Account.find({ ownerId: account.ownerId });
      });
      it('should match snapshot', () => {
        expect(response.text).toMatchSnapshot();
        expect(response.status).toBe(201);
      });
      it('should create a new account in the database', () => {
        expect(dbAccount.length).toBe(1);
      });
    });
    describe('with field missing', () => {
      describe('with ownerId missing', () => {
        it('should match snapshot', async () => {
          const account = {};
          const response = await request(server)
            .post('/account')
            .send(account);
          expect(response.text).toMatchSnapshot();
          expect(response.status).toBe(400);
        });
      });
    });
  });
  describe('PATCH /account', () => {
    let updateData: any;
    let currentAccount: any;
    let response: any;
    let updatedAccount: any;

    beforeAll(async () => {
      const query = { _id: accountId };
      updateData = { balance: 200 };
      currentAccount = await Account.find(query);
      response = await request(server)
        .patch(`/account/${accountId}`)
        .send(updateData);
      updatedAccount = await Account.find(query);
    });
    it('should currently have another balance', () => {
      expect(currentAccount[0].balance !== updateData.balance).toBe(true);
    });
    it('should respond with 204', () => {
      expect(response.status).toBe(204);
    });
    it('should have updated balance in database', () => {
      expect(updatedAccount[0].balance).toBe(updateData.balance);
    });
  });
  describe('DELETE /account/:id', () => {
    let response: any;
    let dbResponse: any;
    beforeAll(async () => {
      const query = { _id: accountId };
      response = await request(server).delete(`/account/${accountId}`);
      dbResponse = await Account.find(query);
    });
    it('should respond with 204', () => {
      expect(response.status).toBe(204);
    });
    it('should be removed from the database', () => {
      expect(dbResponse.length).toBe(0);
    });
  });
});
