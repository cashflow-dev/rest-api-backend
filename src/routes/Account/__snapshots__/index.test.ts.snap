// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/account GET /account should match snapshot 1`] = `"{\\"message\\":\\"ok\\",\\"data\\":[{\\"balance\\":4,\\"name\\":\\"4\\"},{\\"balance\\":3,\\"name\\":\\"3\\"},{\\"balance\\":2,\\"name\\":\\"2\\"},{\\"balance\\":1,\\"name\\":\\"1\\"},{\\"balance\\":0,\\"name\\":\\"0\\"}],\\"next\\":\\"exists\\",\\"statusCode\\":200}"`;

exports[`/account GET /account/:id should match snapshot 1`] = `"{\\"message\\":\\"ok\\",\\"data\\":{},\\"statusCode\\":200}"`;

exports[`/account POST /account with all fields should match snapshot 1`] = `"Created"`;

exports[`/account POST /account with field missing with ownerId missing should match snapshot 1`] = `"{\\"statusCode\\":400,\\"errorCode\\":400,\\"errors\\":[\\"\\\\\\"ownerId\\\\\\" is required\\"],\\"message\\":\\"ValidationError\\"}"`;
